gobblin-metrics/src/main/java/gobblin/metrics/SerializedMetricReportReporter.java

```diff
@@ -0,0 +1,151 @@
+
+  private final Optional<SpecificDatumWriter<MetricReport>> writerOpt;
+
+  public SerializedMetricReportReporter(Builder<?> builder) {
+    super(builder);
+
+    this.byteArrayOutputStream = new ByteArrayOutputStream();
+    this.out = this.closer.register(new DataOutputStream(byteArrayOutputStream));
+    this.encoder = getEncoder(out);
+    this.reportMetrics = this.encoder != null;
+
+    if (this.reportMetrics) {
+      this.writerOpt = Optional.of(new SpecificDatumWriter<MetricReport>(MetricReport.class));
+    } else {
+      this.writerOpt = Optional.absent();
+    }
+
+  }
+
+  public static abstract class Builder<T extends Builder<T>> extends MetricReportReporter.Builder<T> {
+    public Builder(MetricRegistry registry) {
+      super(registry);
+    }
+  }
+
+  @Override
+  protected void pushReport(MetricReport report) {
+    pushSerializedReport(serializeReport(report));
+  }
+
+  /**
+   * Push serialized metric report to metrics sink.
+   * @param serializedReport bytes to send.
+   */
+  protected abstract void pushSerializedReport(byte[] serializedReport);
+
+  /**
+   * Get {@link org.apache.avro.io.Encoder} for serializing Avro records.
+   * @param out {@link java.io.OutputStream} where records should be written.
+   * @return Encoder.
+   */
+  protected Encoder getEncoder(OutputStream out) {
+    try {
+      return EncoderFactory.get().jsonEncoder(MetricReport.SCHEMA$, out);
+    } catch(IOException exception) {
+      LOGGER.warn("KafkaReporter serializer failed to initialize. Will not report Metrics to Kafka.", exception);
+      return null;
+    }
+  }
+
+  /**
+   * Converts a {@link gobblin.metrics.MetricReport} to bytes to send through Kafka.
+   * @param report MetricReport to serialize.
+   * @return Serialized bytes.
+   */
+  protected synchronized byte[] serializeReport(MetricReport report) {
+    if (!this.writerOpt.isPresent()) {
+      return null;
+    }
+
+    try {
+      this.byteArrayOutputStream.reset();
+      // Write version number at the beginning of the message.
+      this.out.writeInt(SCHEMA_VERSION);
+      // Now write the report itself.
+      this.writerOpt.get().write(report, this.encoder);
+      this.encoder.flush();
+      return this.byteArrayOutputStream.toByteArray();
+    } catch(IOException exception) {
+      LOGGER.warn("Could not serialize Avro record for Kafka Metrics. Exception: %s", exception.getMessage());
+      return null;
+    }
+  }
+
+  /**
+   * Parses a {@link gobblin.metrics.MetricReport} from a byte array.
+   * @param reuse MetricReport to reuse.
+   * @param bytes Input bytes.
+   * @return MetricReport.
+   * @throws IOException
+   */
+  public static MetricReport deserializeReport(MetricReport reuse, byte[] bytes) throws IOException {
+    if (!READER.isPresent()) {
+      READER = Optional.of(new SpecificDatumReader<MetricReport>(MetricReport.class));
+    }
+
+    DataInputStream inputStream = new DataInputStream(new ByteArrayInputStream(bytes));
```

Comment:
OK, then please use a Closer to make sure it is properly closed in a finally clause.